cpp/daal/src/data_management/roc_auc_score.cpp

```diff
@@ -0,0 +1,149 @@
+#include "services/daal_defines.h"
+#include "src/algorithms/service_kernel_math.h"
+#include "src/algorithms/service_sort.h"
+#include "src/externals/service_dispatch.h"
+#include "src/externals/service_memory.h"
+#include "src/services/service_data_utils.h"
+#include "src/threading/threading.h"
+
+namespace daal
+{
+namespace data_management
+{
+namespace internal
+{
+template <typename FPType, daal::CpuType cpu>
+services::Status calculateRankDataImpl(FPType * predictedRank, const NumericTablePtr & testPrediction, const size_t nElements)
+{
+    services::Status s;
+    ReadRows<FPType, cpu> testPredictionBlock(testPrediction.get(), 0, 1);
+    const FPType * const testPredictionPtr = testPredictionBlock.get();
+    DAAL_CHECK_BLOCK_STATUS(testPredictionBlock);
+
+    TArrayScalable<IdxValType<FPType>, cpu> predict(nElements);
+    DAAL_CHECK_MALLOC(predict.get());
+
+    const size_t blockSizeDefault = 256;
+    const size_t nBlocks          = nElements / blockSizeDefault + !!(nElements % blockSizeDefault);
+
+    daal::threader_for(nBlocks, nBlocks, [&](const size_t iBlock) {
+        const size_t blockBegin = iBlock * blockSizeDefault;
+        const size_t blockSize  = (iBlock == nBlocks - 1) ? nElements - blockBegin : blockSizeDefault;
+        for (size_t i = 0; i < blockSize; ++i)
+        {
+            predict[blockBegin + i].value = testPredictionPtr[blockBegin + i];
+            predict[blockBegin + i].index = blockBegin + i;
+        }
+    });
+
+    parallel_sort<FPType>(predict.get(), predict.get() + nElements);
+
+    size_t rank            = 1;
+    size_t elementsInBlock = 1;
+    size_t i               = 0;
+    while (i < nElements)
+    {
+        size_t j = i;
+        while ((j < (nElements - 1)) && (predict[j].value == predict[j + 1].value))
+        {
+            j++;
+        }
+        elementsInBlock = j - i + 1;
+
+        PRAGMA_IVDEP
+        PRAGMA_VECTOR_ALWAYS
+        for (size_t j = 0; j < elementsInBlock; ++j)
+        {
+            const size_t idx   = predict[i + j].index;
+            predictedRank[idx] = static_cast<FPType>(rank) + ((static_cast<FPType>(elementsInBlock) - FPType(1.0)) * FPType(0.5));
+        }
+        rank += elementsInBlock;
+        i += elementsInBlock;
+    }
+    return s;
+}
+
+template <typename FPType>
+DAAL_EXPORT void calculateRankData(FPType * predictedRank, const NumericTablePtr & testPrediction, const size_t nElements)
+{
+#define DAAL_CALC_RANK_DATA(cpuId, ...) calculateRankDataImpl<FPType, cpuId>(__VA_ARGS__);
+
+    DAAL_DISPATCH_FUNCTION_BY_CPU_SAFE(DAAL_CALC_RANK_DATA, predictedRank, testPrediction, nElements);
+
+#undef DAAL_CALC_RANK_DATA
+}
+
+template DAAL_EXPORT void calculateRankData<float>(float * predictedRank, const NumericTablePtr & testPrediction, const size_t nElements);
+template DAAL_EXPORT void calculateRankData<double>(double * predictedRank, const NumericTablePtr & testPrediction, const size_t nElements);
```

Comment:
I see no reason to split into two different functions `calculateRankData` and `rocAucScore`. Merge in one:
```python
FPType rocAucScore(const NumericTablePtr & truePrediction, const NumericTablePtr & testPrediction)
```